file_manager.cpp

```diff
@@ -0,0 +1,115 @@
+#include <fcntl.h>
+#include <cstring>
+#include <algorithm>
+#include "file_manager.h"
+#include "file_utils.h"
+#include "file_exceptions.h"
+
+std::string FileManager::getNextLine(int timeout, bool loop) {
+    if (lineCache.empty() || currentLineInBlock >= LINES_IN_BLOCK) {
+        loadLinesToCache(loop);
+    }
+    return lineCache[currentLineInBlock++];
+}
+
+bool FileManager::deleteLine(const std::string &line, int timeout) {
+    lockExclusive();
+    std::string currentLine = loadCurrentLine();
+    if (line != currentLine) {
+        unlock();
+        return false;
+    }
+    deleteCurrentLine();
+    unlock();
+    return true;
+}
+
+void FileManager::writeLine(const std::string &line) {
+    findEmptyLine();
+    lockExclusive();
+    while (loadCurrentLine() != std::string(EMPTY_LINE)) {
+        unlock();
+        findEmptyLine();
+        lockExclusive();
+    }
+    int lineOffset = currentBlock * BLOCK_SIZE + currentLineInBlock * LINE_SIZE;
+    file_utils::writeIn(fd, lineOffset, line.c_str(), LINE_SIZE);
+    unlock();
+}
+
+void FileManager::setFile(const std::string &filepath) {
+    if (file_utils::exists(filepath)) {
+        fd = open(filepath.c_str(), O_RDWR);
+    } else {
+        creat(filepath.c_str(), S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+        fd = open(filepath.c_str(), O_RDWR);
+        fillFileWithEmptyLines();
+    }
+}
+
+void FileManager::loadLinesToCache(bool loop) {
+    if (nextBlocks.empty()) {
+        fillNextBlocksArray();
```

Comment:
Czy nie tutaj powinno być sprawdzenie do loop'a ? Jeśli nextBlocks są puste to znaczy, że właśnie się przeiterowaliśmy po wszystkich chyba ? A pierwszy raz iterować w funkcji setFile.